core/src/main/java/hudson/model/Job.java

```diff
@@ -684,9 +685,37 @@ public void movedTo(DirectlyModifiableTopLevelItemGroup destination, AbstractIte
         }
     }
 
-    @Override public void delete() throws IOException, InterruptedException {
+    @Override
+    public void delete() throws IOException, InterruptedException {
         super.delete();
         Util.deleteRecursive(getBuildDir());
+        if (this instanceof TopLevelItem) {
+            cleanupWorkspaces((TopLevelItem) this);
+        }
+    }
+
+    /**
+     * Clean out workspaces for attached agents (best efforts, as they may not always be online)
+     * There is also hudson.model.WorkspaceCleanupThread to catch ones that were missed.
+     */
+    private void cleanupWorkspaces(TopLevelItem item) {
+        Jenkins j = Jenkins.getInstance();
+        List<Node> nodes = new ArrayList<>();
+        nodes.add(j);
+        nodes.addAll(j.getNodes());
+        Logger log = Logger.getLogger(Job.class.getName());
+        for (Node node : nodes) {
+            FilePath ws = node.getWorkspaceFor(item);
```

Comment:
I have seen Job deletion requests timeout because of how long it took to delete its directory on master. Cleaning workspaces on unpredictable number of nodes definitely needs to be detached from UI thread.